File: apps/api/src/api/services/medication_frequency.py
"""Кратность приёма препарата словами — для отчёта, CSV и PDF (ADR-0033).

Кабинет подписывает кратность своими словарями (`doctor.json`,
`medications.frequencyCodes`), а отчёт собирает сервер, и словарей у него нет.
Две записи одних и тех же слов сверяет тест
(`test_medication_frequency.py`): семья и врач не должны
читать в отчёте одно, а в карте другое.
"""

from __future__ import annotations

from core.models.enums import MedicationFrequency

LABELS: dict[MedicationFrequency, str] = {
    MedicationFrequency.ONCE_DAILY: "1 раз в сутки",
    MedicationFrequency.TWICE_DAILY: "2 раза в сутки",
    MedicationFrequency.THREE_TIMES_DAILY: "3 раза в сутки",
    MedicationFrequency.FOUR_TIMES_DAILY: "4 раза в сутки",
    MedicationFrequency.EVERY_OTHER_DAY: "Через день",
    MedicationFrequency.AS_NEEDED: "По требованию",
    MedicationFrequency.OTHER: "Другая схема",
}


def describe_frequency(code: MedicationFrequency | None, note: str | None) -> str:
    """«2 раза в сутки — утром и на ночь».

    У «другой схемы» сама подпись ничего не сообщает, поэтому печатается одно
    уточнение. У записи до списка кода нет, и кратность целиком — в уточнении.
    """

    if code is None or code is MedicationFrequency.OTHER:
        return note or ""
    label = LABELS[code]
    return f"{label} — {note}" if note else label

File: apps/web/src/features/doctor/medicationFrequency.ts
import type { Medication } from "./types";

export type MedicationFrequency = NonNullable<Medication["frequency_code"]>;

/**
 * Кратность приёма — как в перечислении сервера (`MedicationFrequency`,
 * ADR-0033). Порядок — порядок списка в форме: от редкого к частому, затем
 * то, что расписанием не описывается.
 *
 * Полноту списка держит связка тестов: `medicationFrequency.test.ts` сверяет его
 * со словарём, а тест API — словарь с подписями сервера.
 */
export const MEDICATION_FREQUENCIES = [
  "once_daily",
  "twice_daily",
  "three_times_daily",
  "four_times_daily",
  "every_other_day",
  "as_needed",
  "other",
] as const satisfies readonly MedicationFrequency[];

export function isMedicationFrequency(
  value: string,
): value is MedicationFrequency {
  return (MEDICATION_FREQUENCIES as readonly string[]).includes(value);
}

/**
 * «2 раза в сутки — утром и на ночь» — тем же правилом, что и отчёт сервера
 * (`api/services/medication_frequency.py`).
 *
 * У «другой схемы» подпись ничего не сообщает, поэтому показывается одно
 * уточнение. У записи, заведённой до списка, кода нет, и кратность целиком в
 * словах.
 */
export function describeFrequency(
  medication: Pick<Medication, "frequency_code" | "frequency">,
  labelOf: (code: MedicationFrequency) => string,
): string {
  const note = medication.frequency ?? "";
  const code = medication.frequency_code;
  if (code === null || code === undefined || code === "other") return note;
  return note === "" ? labelOf(code) : `${labelOf(code)} — ${note}`;
}
